<a href="https://colab.research.google.com/github/toejo/credit-risk-classification/blob/main/credit_risk_classification.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
# Import the modules
import numpy as np
import pandas as pd
from sklearn.metrics import balanced_accuracy_score, confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [2]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
df = pd.read_csv('lending_data.csv')

# Review the DataFrame
df.sample(10)

loan_size  interest_rate  borrower_income  debt_to_income  \
14691    11600.0          8.049            56300        0.467140   
38612    10300.0          7.504            51200        0.414062   
74830     6900.0          6.062            37600        0.202128   
18974     9800.0          7.288            49200        0.390244   
20853     9600.0          7.222            48600        0.382716   
33085    11400.0          7.989            55800        0.462366   
9237      8700.0          6.825            44800        0.330357   
57320     7800.0          6.431            41100        0.270073   
28904     9400.0          7.101            47400        0.367089   
54567    10800.0          7.699            53100        0.435028   

       num_of_accounts  derogatory_marks  total_debt  loan_status  
14691                5                 1       26300            0  
38612                4                 1       21200            0  
74830                1                 0        7600            0  
18974                4                 0       19200            0  
20853                4                 0       18600            0  
33085                5                 1       25800            0  
9237                 3                 0       14800            0  
57320                2                 0       11100            0  
28904                3                 0       17400            0  
54567                5                 1       23100            0

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [3]:
# Separate the data into labels and features

# Separate the y variable, the labels
y = df['loan_status']

# Separate the X variable, the features
# df.columns ---> to get the column names
X = df[['loan_size', 'interest_rate', 'borrower_income', 'debt_to_income',
       'num_of_accounts', 'derogatory_marks', 'total_debt']]

In [4]:
# Review the y variable Series
y.sample(10)

59269    0
25631    0
2743     0
34684    0
67999    0
21026    0
73797    0
37693    0
19597    0
60755    0
Name: loan_status, dtype: int64

In [5]:
# Review the X variable DataFrame
X.sample(10)

loan_size  interest_rate  borrower_income  debt_to_income  \
53803     9500.0          7.152            47900        0.373695   
67028     7700.0          6.405            40900        0.266504   
19969    10200.0          7.455            50800        0.409449   
18636     9900.0          7.319            49500        0.393939   
5063     10700.0          7.657            52700        0.430740   
36954     7700.0          6.397            40800        0.264706   
65258    11000.0          7.819            54200        0.446494   
10549     8600.0          6.783            44400        0.324324   
37948     9400.0          7.129            47700        0.371069   
41012     9200.0          7.044            46900        0.360341   

       num_of_accounts  derogatory_marks  total_debt  
53803                4                 0       17900  
67028                2                 0       10900  
19969                4                 1       20800  
18636                4                 0       19500  
5063                 5                 1       22700  
36954                2                 0       10800  
65258                5                 1       24200  
10549                3                 0       14400  
37948                3                 0       17700  
41012                3                 0       16900

### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [6]:
# Check the balance of our target values
y.value_counts()

0    75036
1     2500
Name: loan_status, dtype: int64

### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [7]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state = 1, stratify = y, train_size = 0.7)


---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [8]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
logreg_model = LogisticRegression(random_state = 1)


# Fit the model using training data
logreg_model.fit(X_train, y_train)

LogisticRegression(random_state=1)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [9]:
# Make a prediction using the testing data
y_pred = logreg_model.predict(X_test)

pred_results = pd.DataFrame({
    "test": y_test,
    "pred": y_pred
})

pred_results.sample(10)

test  pred
64825     0     0
27938     0     0
37391     0     0
48092     0     0
63750     0     0
54729     0     0
2791      0     0
30006     0     0
5122      0     0
24913     0     0

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [10]:
# Print the balanced_accuracy score of the model
acc_score = balanced_accuracy_score(y_test, y_pred)
print(f"The balanced accuracy score is {round(acc_score,2)}.")

The balanced accuracy score is 0.95.


In [11]:
# Generate a confusion matrix for the model
conf_matr = confusion_matrix(y_test, y_pred)
print("Confusion matrix:")
print(conf_matr)

Confusion matrix:
[[22409   102]
 [   66   684]]


In [12]:
# Print the classification report for the model
clasf_repr = classification_report(y_test, y_pred)
print("Classification report:")
print(clasf_repr)

Classification report:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00     22511
           1       0.87      0.91      0.89       750

    accuracy                           0.99     23261
   macro avg       0.93      0.95      0.94     23261
weighted avg       0.99      0.99      0.99     23261



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** The logistic regression model predicted both at a 99% acuracy. Arguably, the model did a better job at predicting healthy loans/0 (both recall and f1-score at 1) compared to its performance when predicting high-risk loans/1 (0.91 and 0.89 recall and f1-score, respectively).

---

## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points.

In [15]:
# Import the RandomOverSampler module form imbalanced-learn
from imblearn.over_sampling import RandomOverSampler

# Instantiate the random oversampler model
# # Assign a random_state parameter of 1 to the model
ranovsam_model = RandomOverSampler(random_state = 1)

# Fit the original training data to the random_oversampler model
X_resampled, y_resampled = ranovsam_model.fit_resample(X_train, y_train)

In [16]:
# Count the distinct values of the resampled labels data
y_resampled.value_counts()

0    52525
1    52525
Name: loan_status, dtype: int64

### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [23]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
resampled_model = LogisticRegression(random_state = 1)

# Fit the model using the resampled training data
resampled_model.fit(X_resampled, y_resampled)

# Make a prediction using the testing data
resampled_y_pred = resampled_model.predict(X_test)

resampled_results = pd.DataFrame({
    "test": y_test,
    "pred": resampled_y_pred
})
resampled_results.sample(20)

test  pred
23847     0     0
68637     0     0
43461     0     0
74851     0     0
75809     1     1
58748     0     0
9593      0     0
77387     1     1
76015     1     1
6936      0     0
38638     0     0
43154     0     0
56853     0     0
40369     0     0
23348     0     0
40026     0     0
23853     0     0
73983     0     0
73224     0     0
59139     0     0

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [28]:
# Print the balanced_accuracy score of the model
resampled_acc_score = balanced_accuracy_score(y_test, resampled_y_pred)
print(f"The balanced accuracy score is {round(resampled_acc_score, 3)}.")

The balanced accuracy score is 0.996.


In [29]:
# Generate a confusion matrix for the model
resampled_conf_matr = confusion_matrix(y_test, resampled_y_pred)

print("Confusion matrix:")
print(resampled_conf_matr)

Confusion matrix:
[[22397   114]
 [    2   748]]


In [30]:
# Print the classification report for the model
resampled_clasf_repr = classification_report(y_test, resampled_y_pred)

print("Classification report:")
print(resampled_clasf_repr)

Classification report:
              precision    recall  f1-score   support

           0       1.00      0.99      1.00     22511
           1       0.87      1.00      0.93       750

    accuracy                           1.00     23261
   macro avg       0.93      1.00      0.96     23261
weighted avg       1.00      1.00      1.00     23261



### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** The logistic regression model (model B) that used an oversampled data resulted in an accuracy of 1.0, whereas the non-oversampled logistic regression model (model A) resulted in an accuracy of 0.99. The accuracy results are comparable.

The difference is in the recall and f1-score value of the healthy/0 and high-risk/1 loan. For healthy/0 loans, model B yielded comparable values (recall = 0.99 and f1-score = 1.0 ) with model A (recall = 1.0 and f1-score = 1.0). However, model B yielded better values (recall = 1.0 and f1-score = 0.93) for high-risk/1 loans as compared to model A (recall = 0.91 and f1-score = 0.89). Therefore, model B did a better job at predicting loan status.